### DISCUSSION TITLE
Schema.org annotation for the rendered course

### DISCUSSION THREAD

**Discussion by abotzki** [LiaScript/LiaScript#150] [Ideas]
Hi @andre-dietrich,

related to this discussion #118, I was wondering whether the macro `script` could be also used to digest a JSON-LD object and inject it in the header of the rendered Liascript course.
If it could be ingested via such a mechanism, the rendered course via https://liascript.github.io/?https://raw.... would contain the complete JSON-LD object.

Below, you'll find an example of a manually created entry. 

```
<script type="application/ld+json">
  {
    "@context": "https://schema.org/",
    "@type": "LearningResource",
    "@id": "https://elixir-europe-training.github.io/ELIXIR-TrP-TeSS/",
    "http://purl.org/dc/terms/conformsTo": {
        "@type": "CreativeWork",
        "@id": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE"
    },
    "description": "TeSS, how can I help you? This is our interactive hands-on course about efficient use of the ELIXIR TeSS platform.",
    "keywords": "FAIR, OPEN, Bioinformatics, Teaching, TeSS",
    "name": "TeSS, how can I help you?",
    "license": "https://creativecommons.org/licenses/by/4.0/",
    "educationalLevel": "beginner",
    "competencyRequired": "none",
    "teaches": ["search events and material in TeSS via direct and faceted search", "add manually and automatically events and material to TeSS", "extract events and material from TeSS by using TeSS widgets"],
    "audience": "training providers",
    "inLanguage": "en-US",
    "learningResourceType": ["tutorial"],
    "author":  [
    {
    "@type": "Person",
    "name": "Bruna Piereck"},
    { 
    "@type": "Person",
    "name": "Olivier Sand"},
    { 
    "@type": "Person",
    "name": "Alexander Botzki"
    }],
    "contributor":  [
    {
    "@type": "Person",
    "name": "Yasmine Maes"},
    { 
    "@type": "Person",
    "name": "Finn Bacall"},
    { 
    "@type": "Person",
    "name": "Munazah Andrabi"
    }]
  }
</script>
```

**Comment by andre-dietrich** [comment]
@abotzki ... this is actually possible, you can try this out [here](https://liascript.github.io/LiveEditor/?/show/code/H4sIAAAAAAAAA51VXW8bRRR9319xMQ9JhXcdJJ5KW5K0qRTkNlESVBBC8nj32jtkdmaZmU3iov4LJB75i/wEzp1ZG1MqJCJZtnfnfpx7zr13nn1WlsXxt9cXb+evimeh9rqP5AdbOlvzi4LIcKSfg7P0nI6PqMCb2tkQaTR9To2rh45trGrPKvKZYXk6PMgGB0++hkv+X8VNz/A4UH1vdK2idnZmmi8k/MG+GT9IYAFVhei1XevV5lCsEAxmu4wtq6ZCMLbNy1ab5jAHyFazGcVWB8LHWbMhy9xwQ9FR7/kO7qRsQ26I/RCn2V4FMs6u5Te2TJ7DYCK5FaltuQg22IZX2nIzOun01jpQ0rp7S87DETYeye41IFiaa3Wd/Ef2nOGq4eWwPpyoRkAlgoFMCpo8KZ7NcroXxTEiFcUxCnCqKZRhHw8nNxevLp6StgFPtHGDz/4tUsI5uZTli6L4HIntLcK/UlHRn3/89ntRnFhiw2uF8gFjEAloBcjAAZ7JZIdGHJqNVR1kMiCv0+s20pITL0NQaxZanMWPZ+rkqx5CdB11qvYuVPQuly72UpWYJzm24skJvJI6/3QmSQkQi7EtF1O6b3XdjlqFnRplVOtEccrSK686juxpse3fRZbY7h2WYViGqGOufHF8tKjoPMPkB9X1hlGlcfdIUruGy6Vx9a2gHEXP+VRGWgpSCIEeCKnKim6kxo6VlRZSMQe2UaPUpWs2mQb+KHavQkDose3eMziAJCvtMWc75BXEQ+2WlMGTxfTc8VTkB1/yMjgRMUX4Vt2NDZcSASF6Sw4WYyctEpgRrU6M9lxrZUYFEnR0alToMjRKlxHn99uBkkAs9NbA0Qw+9Y/bSs1/j2kF6IvFIjUp0XbZ/IphmBxLEoz75ClN2hj78HSG3m+5U5Xz69lkmoxkb4jFnJW3SHPFAV1f83ism313NvpB+5IH73ouo0cJcKnW0G1YVtrNzubn359flTf+srzh6+sxibjDux+8SambegaauzADQgxIF24csgjofUQvZedBiHfO36Y4/8az1C5XFFLc3ruVNhxmNyOyN1iaHtTPvqyOyquz+dnJ9dkEkT4kWA3nXke3SlABPCXsmST6OZrO9NID3+yklHUAvdmrWoBRixEImAYIOPjApJYYI+LVStdaxnAIvBUsE0OSg3qjopRdZXZueXPvfBMEw+uT86spXVyevZ3SqXY60QMW6jCFr6ox9utpipJ9sUX4P7FnM9wIjJW2T109slu7rsPaTPyNZmG23My+qo5G9bgZ8n2izBzL3UiUJa+1teyzBWL0GCNbb674lwHzmESyWA75PApylgJ/zDIGNBuWTroqQlok3aiUTHwi6U5jSyJUne+SlaqRoaHsuW0HGcpO2SHtUTFTA1ad8JXefCo+pnVHnxT3EEXNT5quPNZmArPcQEuZwfR0r5s1xyCN9FOqTw0N9K4Tv9uhwE3o7jSuqpA50Hau7HrAdhcrtuV3o4Tmo8G7ye0/MoVl6gTMfrLYOr+zyFOzPzeXyImOnm4Ptj1yitWt6FLjaqhvJ+n0w/R/xrgw+g4B6BpEPTLEicFlIHc4nbr4/laPYXb1ydbyeil1P6LIH1TocJfQG8XhkQBfo7PpVEkLPTLCGxD9XrV0YhuvlvsVFh9kW/8F3gKeuRkKAAA=)
...

You can use a script-tag for this, for which `@JSONLD` is a simple substitution or you put the same code with your Json into the onload-definition. Using a script/macro will only be executed if the slide is opened, the onload-part will once if the course is loaded ...

```` markdown
<!--
@JSONLD
<script run-once>
  let json = @0 

  const script = document.createElement('script');
  script.type = 'application/ld+json';
  script.text = JSON.stringify(json);

  document.head.appendChild(script);

  // this is only needed to prevent and output,
  // as long as the result of a script is undefined,
  // it is not shown or rendered within LiaScript
  console.debug("added json to head")
</script>
@end

@onload
alert("TODO: insert your json here")
@end
-->

# Linked Data 🔗

An elegant solution for adding linked data dynamically might be the usage of one ore more custom macros.
Within the head of this document there is one custom macro called `@JSONLD`, which outputs a script-tag
with the parameter `run-once` and one parameter-substitution `@0`. In the example below a code-block is
defined with a macro-call in its head. This means that the entire body of the code-block is passed as the
zero or first parameter.

As an alternative, you can also add the JavaScript code into the `@onload` body. This is a special macro
that contains some code that is only loaded once, during load of the document...

```json   @JSONLD
{
  "@context": "https://schema.org/",
  "@type": "LearningResource",
  "@id": "https://elixir-europe-training.github.io/ELIXIR-TrP-TeSS/",
  "http://purl.org/dc/terms/conformsTo": {
    "@type": "CreativeWork",
    "@id": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE"
  },
  "description": "TeSS, how can I help you? This is our interactive hands-on course about efficient use of the ELIXIR TeSS platform.",
  "keywords": "FAIR, OPEN, Bioinformatics, Teaching, TeSS",
  "name": "TeSS, how can I help you?",
  "license": "https://creativecommons.org/licenses/by/4.0/",
  "educationalLevel": "beginner",
  "competencyRequired": "none",
  "teaches": [
    "search events and material in TeSS via direct and faceted search",
    "add manually and automatically events and material to TeSS",
    "extract events and material from TeSS by using TeSS widgets"
  ],
  "audience": "training providers",
  "inLanguage": "en-US",
  "learningResourceType": [
    "tutorial"
  ],
  "author": [
    {
      "@type": "Person",
      "name": "Bruna Piereck"
    },
    {
      "@type": "Person",
      "name": "Olivier Sand"
    },
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Yasmine Maes"
    },
    {
      "@type": "Person",
      "name": "Finn Bacall"
    },
    {
      "@type": "Person",
      "name": "Munazah Andrabi"
    }
  ]
}
```
````
Here is a screenshot of the example with the added JSON-LD part ...

![Bildschirmfoto vom 2024-01-03 15-18-02](https://github.com/LiaScript/LiaScript/assets/3089101/aca535b2-a7dc-4862-bf46-ff84b083e1d8)
